File: 백준/Gold/13023. ABCDE/ABCDE.cc
#include<bits/stdc++.h>
using namespace std;
bool visited[2002];
int n, m, a, b, result;
vector<int> v[2002];

void dfs(int node, int depth){
  if(depth == 4){
    result = 1;
    return;
  }
  visited[node] = 1;
  for(int i : v[node]){
    if(!visited[i]) dfs(i, depth + 1);
  }
  visited[node] = 0;
}
int main(){
  cin >> n >> m;
  for(int i = 0; i < m; i++){
    cin >> a >> b;
    v[a].push_back(b);
    v[b].push_back(a);
  }
  for(int i = 0; i < n; i++){
    dfs(i, 0);
    if(result) break;
  }
  cout << result << '\n';
  return 0;
  
}

File: 백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static LinkedList<Integer>[] graph;
	static boolean[] visited;
	static int N, M;
	static boolean result = false;
	public static void main(String[] args) throws IOException {
		String[] input = br.readLine().split(" ");
		N = Integer.parseInt(input[0]);
		M = Integer.parseInt(input[1]);
		
		graph = new LinkedList[N];
		
		for(int i = 0; i < N; i++) {
			graph[i] = new LinkedList<Integer>();
		}
		
		
		for(int i = 0; i < M; i++) {
			String[] nodes = br.readLine().split(" ");
			int node1 = Integer.parseInt(nodes[0]);
			int node2 = Integer.parseInt(nodes[1]);
			graph[node1].add(node2);
			graph[node2].add(node1);
		}
		
		for(int i = 0; i < N; i++) {
			visited = new boolean[N];
			dfs(i, 1); 
            if(result){
				System.out.println(1);
				return;
			}
		}
		
		System.out.println(0);
	}
	
	static void dfs(int start, int count) {
        if(count == 5) {
			result = true;
            return;
		}
		visited[start] = true;
		for(int nextNode: graph[start]) {
			if(!visited[nextNode]) {
				dfs(nextNode, count + 1);
			}
		}
		visited[start] = false;
	}
}